### PR TITLE
add PascalsTriangle generator

### LINKED ISSUE #418: Add Pascals Triangle Test Generator
This issue is part of an overall initiative to complete #195.

The documentation on how to add generators can be found [here](https://github.com/exercism/csharp/blob/master/docs/GENERATORS.md). If you get stuck or have any questions about adding this generator, please do not hesitate to reach out!

### PR DESCRIPTION
**Pull request by bmeverett**
Adds the PascalsTriangle generator. Pretty straight forward, except for the exception cases. I modified the Example to get around the yield return so the exception would be thrown. And the final test case, where the value is null,  I converted that back to -1 because you can't run a test with the incorrect number of parameters as the solution doesn't build. 

Closes #418

### CODE REVIEW

File: generators/Exercises/PascalsTriangle.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        {
+            foreach (var dataCases in canonicalData.Cases)
+            {
+                dataCases.UseVariableForExpected = true;
+                dataCases.Property = "calculate";
+                if (!(dataCases.Expected is JArray))
+                    dataCases.ExceptionThrown = typeof(ArgumentOutOfRangeException);
+                if (dataCases.Properties["count"] == null)
+                    dataCases.Properties["count"] = -1;
+            }
```
Comment (jpreese) [inline generators/Exercises/PascalsTriangle.cs]: Should we be converting this to a -1? It should be entirely possible to pass in null (int?), or pass in nothing with a default param.

File: exercises/pascals-triangle/Example.cs [hunk 2]
```diff
@@ -1,15 +1,24 @@
...
     public static IEnumerable<IEnumerable<int>> Calculate(int rows)
     {
+        if (rows < 0)
+            throw new ArgumentOutOfRangeException();
+
+        return ItterateRows(rows);
+    }
+    
+    private static IEnumerable<IEnumerable<int>> ItterateRows(int rows)
+    {
```
Comment (jpreese) [inline exercises/pascals-triangle/Example.cs]: ItterateRows should be IterateRows

File: generators/Exercises/PascalsTriangle.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+using Newtonsoft.Json.Linq;
+
+namespace Generators.Exercises
+{
+    public class PascalsTriangle : Exercise
+    {
+        protected override void UpdateCanonicalData(CanonicalData canonicalData)
+        {
+            foreach (var dataCases in canonicalData.Cases)
+            {
```
Comment (jpreese) [inline generators/Exercises/PascalsTriangle.cs]: I would change dataCases to dataCase (most use canonicalDataCase). Since the variable is a single case.

**Comment by jpreese** [comment]
Just some small nits @bmeverett, looks good though!

File: generators/Exercises/PascalsTriangle.cs (hunk 1 above)
Comment (ErikSchierboom) [reply generators/Exercises/PascalsTriangle.cs]: @jpreese Well, I think the canonical data is at fault here. In what way would passing null make sense for calculating pascal's triangle? I've created [an issue](https://github.com/exercism/problem-specifications/issues/925) to discuss removing this strange case. In the mean time, I'm in favor of manually removing this error case from the canonical data to make sure we don't output it.

Comment (ErikSchierboom) [reply generators/Exercises/PascalsTriangle.cs]: @jpreese  @bmeverett I've created a [PR](https://github.com/exercism/problem-specifications/pull/926) to fix the canonical data.

File: exercises/pascals-triangle/PascalsTriangle.cs [hunk 4]
```diff
@@ -7,4 +7,14 @@ public static IEnumerable<IEnumerable<int>> Calculate(int rows)
     {
         throw new NotImplementedException();
     }
+
+    private static IEnumerable<int> Row(int row)
```
Comment (ErikSchierboom) [inline exercises/pascals-triangle/PascalsTriangle.cs]: Could you remove this function? The stub implementation only needs the public methods to allow it to compile.

File: exercises/pascals-triangle/PascalsTriangle.cs [hunk 5]
```diff
@@ -7,4 +7,14 @@ public static IEnumerable<IEnumerable<int>> Calculate(int rows)
...
+    {
+        throw new NotImplementedException();
+    }
+
+    private static IEnumerable<IEnumerable<int>> IterateRows(int rows)
```
Comment (ErikSchierboom) [inline exercises/pascals-triangle/PascalsTriangle.cs]: Could you remove this function? The stub implementation only needs the public methods to allow it to compile.

**Comment by ErikSchierboom** [comment]
Merged. Thanks a lot! :tada:

File: generators/Exercises/PascalsTriangle.cs (hunk 1 above)
Comment (jpreese) [reply generators/Exercises/PascalsTriangle.cs]: Yeah, I'd agree. I know there was a PR over in the problem-specs repo about wanting to test edge cases (e.g. null). I'm not sure if this is a result of that or where that got left off.
